morphir/runtime/src/org/finos/morphir/runtime/ToMDMConcept.scala

```diff
@@ -0,0 +1,165 @@
+package org.finos.morphir.runtime
+
+import org.finos.morphir.ir.conversion
+
+import org.finos.morphir.naming._
+import org.finos.morphir.datamodel.{Concept, Label, EnumLabel}
+import org.finos.morphir.ir.AccessControlled
+import org.finos.morphir.ir.{Type => T}
+import org.finos.morphir.ir.Type.{Type, UType}
+
+import scala.collection.immutable.{Map, Set}
+import org.finos.morphir.runtime.Distributions
+import org.finos.morphir.datamodel.Concept.Defaults
+import org.finos.morphir.runtime.Extractors.Types.*
+
+trait ToMDMConcept[A] { self =>
+  import ToMDMConcept.*
+  def apply(distributions: Distributions, bindings: Map[Name, Concept] = Map.empty): ToMDMConceptEither
+
+  final def concept(distributions: Distributions, bindings: Map[Name, Concept] = Map.empty): ToMDMConceptEither =
+    apply(distributions, bindings)
+
+  def as[B]: ToMDMConcept[B] = new ToMDMConcept[B] {
+    override def apply(distributions: Distributions, bindings: Map[Name, Concept] = Map.empty): ToMDMConceptEither =
+      self.apply(distributions, bindings)
+  }
+}
+
+object ToMDMConcept {
+
+  sealed trait NoEquivalentMDM         extends Throwable
+  case class Unsupported(msg: String)  extends NoEquivalentMDM
+  case class MissingAlias(msg: String) extends NoEquivalentMDM
+  type ToMDMConceptEither = Either[NoEquivalentMDM, Concept]
+
+  def apply[A](implicit toMDMConcept: ToMDMConcept[A]): ToMDMConcept[A] = toMDMConcept
+  // def summon[A]: SummonPartiallyApplied[A]                              = new SummonPartiallyApplied[A]
+
+  def toConceptConverter[A](f: => Concept): ToMDMConcept[A] = new ToMDMConcept[A] {
+    def apply(distributions: Distributions, bindings: Map[Name, Concept] = Map.empty): ToMDMConceptEither = Right(f)
+  }
+  def toFailedConverter[A](f: => NoEquivalentMDM): ToMDMConcept[A] = new ToMDMConcept[A] {
+    def apply(distributions: Distributions, bindings: Map[Name, Concept] = Map.empty): ToMDMConceptEither = Left(f)
+  }
+
+  implicit def uTypeToConcept(tpe: UType): ToMDMConcept[UType] =
+    tpe match {
+      case _: Type.ExtensibleRecord[_] => toFailedConverter(Unsupported("Extensible record type has no MDM equivalent"))
```

Comment:
Why not make these more concrete errors... while `Unsupported` works I feel going more granular here is more helpful to the user.